File: app/src/main/java/com/zacharee1/systemuituner/fragmenthelpers/StatbarHelper.java
package com.zacharee1.systemuituner.fragmenthelpers;

import android.preference.Preference;
import android.preference.SwitchPreference;
import android.provider.Settings;

import com.zacharee1.systemuituner.fragments.ItemDetailFragment;
import com.zacharee1.systemuituner.util.SettingsUtils;

public class StatbarHelper extends BaseHelper
{
    public static final String RESET_BLACKLIST = "reset_blacklist";
    public static final String BACKUP_BLACKLIST = "backup_blacklist";
    public static final String RESTORE_BLACKLIST = "restore_blacklist";
    public static final String ICON_BLACKLIST = "icon_blacklist";
    public static final String ICON_BLACKLIST_BACKUP = "icon_blacklist_backup";

    public StatbarHelper(ItemDetailFragment fragment) {
        super(fragment);

        preferenceListeners();
        setSwitchPreferenceStates();
        switchPreferenceListeners();
    }

    private void preferenceListeners() {
        Preference resetBL = findPreference(RESET_BLACKLIST);
        Preference backupBL = findPreference(BACKUP_BLACKLIST);
        Preference restoreBL = findPreference(RESTORE_BLACKLIST);

        resetBL.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener()
        {
            @Override
            public boolean onPreferenceClick(Preference preference)
            {
                SettingsUtils.writeSecure(getContext(), ICON_BLACKLIST, "");
                setSwitchPreferenceStates();
                return true;
            }
        });

        backupBL.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener()
        {
            @Override
            public boolean onPreferenceClick(Preference preference)
            {
                String currentBL = Settings.Secure.getString(getContext().getContentResolver(), ICON_BLACKLIST);
                SettingsUtils.writeGlobal(getContext(), ICON_BLACKLIST_BACKUP, currentBL);
                setSwitchPreferenceStates();
                return true;
            }
        });

        restoreBL.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener()
        {
            @Override
            public boolean onPreferenceClick(Preference preference)
            {
                String backupBL = Settings.Global.getString(getContext().getContentResolver(), ICON_BLACKLIST_BACKUP);
                SettingsUtils.writeSecure(getContext(), ICON_BLACKLIST, backupBL);
                setSwitchPreferenceStates();
                return true;
            }
        });
    }

    private void setSwitchPreferenceStates() {
        SettingsUtils.shouldSetSwitchChecked(getFragment());
    }

    private void switchPreferenceListeners() {
        for (int i = 0; i < getPreferenceScreen().getRootAdapter().getCount(); i++) {
            Object o = getPreferenceScreen().getRootAdapter().getItem(i);

            if (o instanceof SwitchPreference) {
                final SwitchPreference pref = (SwitchPreference) o;

                pref.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener()
                {
                    @Override
                    public boolean onPreferenceChange(Preference preference, Object o)
                    {
                        String key = preference.getKey();
                        boolean value = Boolean.valueOf(o.toString());

                        SettingsUtils.changeBlacklist(key, value, getContext());
                        return true;
                    }
                });
            }
        }
    }

    @Override
    public void onDestroy() {

    }
}

File: app/src/main/java/com/zacharee1/systemuituner/fragmenthelpers/StatbarHelper.kt
package com.zacharee1.systemuituner.fragmenthelpers

import android.content.Intent
import android.os.Bundle
import android.preference.Preference
import android.preference.SwitchPreference
import android.provider.Settings
import android.view.View
import com.zacharee1.systemuituner.R
import com.zacharee1.systemuituner.activites.ItemDetailActivity

import com.zacharee1.systemuituner.fragments.ItemDetailFragment
import com.zacharee1.systemuituner.util.SettingsUtils

class StatbarHelper(fragment: ItemDetailFragment) : BaseHelper(fragment) {

    init {
        preferenceListeners()
        setSwitchPreferenceStates()
        switchPreferenceListeners()
    }

    private fun preferenceListeners() {
        val resetBL = findPreference(RESET_BLACKLIST)
        val backupBL = findPreference(BACKUP_BLACKLIST)
        val restoreBL = findPreference(RESTORE_BLACKLIST)
        val auto = findPreference(AUTO_DETECT)

        resetBL?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            SettingsUtils.writeSecure(context, ICON_BLACKLIST, "")
            setSwitchPreferenceStates()
            true
        }

        backupBL?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val currentBL = Settings.Secure.getString(context.contentResolver, ICON_BLACKLIST)
            SettingsUtils.writeGlobal(context, ICON_BLACKLIST_BACKUP, currentBL)
            setSwitchPreferenceStates()
            true
        }

        restoreBL?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val backupBLString = Settings.Global.getString(context.contentResolver, ICON_BLACKLIST_BACKUP)
            SettingsUtils.writeSecure(context, ICON_BLACKLIST, backupBLString)
            setSwitchPreferenceStates()
            true
        }

        auto?.setOnPreferenceClickListener {
            if (fragment.activity.findViewById<View>(R.id.item_detail_container) != null) {
                val arguments = Bundle()
                arguments.putString(ItemDetailFragment.ARG_ITEM_ID, "auto")
                val fragment = ItemDetailFragment()
                fragment.arguments = arguments
                activity
                        ?.fragmentManager
                        ?.beginTransaction()
                        ?.replace(R.id.item_detail_container, fragment)
                        ?.commit()
            } else {
                val intent = Intent(context, ItemDetailActivity::class.java)
                intent.putExtra(ItemDetailFragment.ARG_ITEM_ID, "auto")

                context.startActivity(intent)
            }
            true
        }
    }

    private fun setSwitchPreferenceStates() {
        SettingsUtils.shouldSetSwitchChecked(fragment)
    }

    private fun switchPreferenceListeners() {
        (0 until preferenceScreen.rootAdapter.count)
                .map { preferenceScreen.rootAdapter.getItem(it) }
                .filterIsInstance<SwitchPreference>()
                .forEach {
                    it.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { preference, o ->
                        val key = preference.key
                        val value = java.lang.Boolean.valueOf(o.toString())

                        SettingsUtils.changeBlacklist(key, value, context)
                        true
                    }
                }
    }

    override fun onDestroy() {

    }

    companion object {
        const val RESET_BLACKLIST = "reset_blacklist"
        const val BACKUP_BLACKLIST = "backup_blacklist"
        const val RESTORE_BLACKLIST = "restore_blacklist"
        const val ICON_BLACKLIST = "icon_blacklist"
        const val ICON_BLACKLIST_BACKUP = "icon_blacklist_backup"
        const val AUTO_DETECT = "auto_detect"
    }
}
